metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java

```diff
@@ -527,227 +481,192 @@ public ListResult<RecordTemplate> listLatestAspects(
+      // Retention optimization and tx
+      if (_retentionService != null) {
+        List<RetentionService.RetentionContext> retentionBatch = upsertResults.stream()
+                // Only consider retention when there was a previous version
+                .filter(result -> latestAspects.containsKey(result.getUrn().toString())
+                        && latestAspects.get(result.getUrn().toString()).containsKey(result.getRequest().getAspectName()))
+                .filter(result -> {
+                  RecordTemplate oldAspect = result.getOldValue();
+                  RecordTemplate newAspect = result.getNewValue();
+                  // Apply retention policies if there was an update to existing aspect value
+                  return oldAspect != newAspect && oldAspect != null && _retentionService != null;
+                })
+                .map(result -> RetentionService.RetentionContext.builder()
+                        .urn(result.getUrn())
+                        .aspectName(result.getRequest().getAspectName())
+                        .maxVersion(Optional.of(result.getMaxVersion()))
+                        .build())
+                .collect(Collectors.toList());
+        _retentionService.applyRetentionWithPolicyDefaults(retentionBatch);
+      } else {
+        log.warn("Retention service is missing!");
       }
-      return result;
-    }, DEFAULT_MAX_TRANSACTION_RETRY);
+
+      return upsertResults;
+    }, aspectsBatch, DEFAULT_MAX_TRANSACTION_RETRY);
   }
 
   @Nonnull
-  protected SystemMetadata generateSystemMetadataIfEmpty(@Nullable SystemMetadata systemMetadata) {
-    if (systemMetadata == null) {
-      systemMetadata = new SystemMetadata();
-      systemMetadata.setRunId(DEFAULT_RUN_ID);
-      systemMetadata.setLastObserved(System.currentTimeMillis());
+  private List<UpdateAspectResult> emitMCL(List<UpdateAspectResult> sqlResults, boolean emitMCL) {
+    List<UpdateAspectResult> withEmitMCL = sqlResults.stream()
+            .map(result -> emitMCL ? conditionallyProduceMCLAsync(result) : result)
+            .collect(Collectors.toList());
+
+    // join futures messages, capture error state
+    List<Pair<Boolean, UpdateAspectResult>> statusPairs = withEmitMCL.stream()
+            .filter(result -> result.getMclFuture() != null)
+            .map(result -> {
+              try {
+                result.getMclFuture().get();
+                return Pair.of(true, result);
+              } catch (InterruptedException | ExecutionException e) {
+                return Pair.of(false, result);
+              }
+            }).collect(Collectors.toList());
+
+    if (statusPairs.stream().anyMatch(p -> !p.getFirst())) {
+      log.error("Failed to produce MCLs: {}", statusPairs.stream()
+              .filter(p -> !p.getFirst())
+              .map(Pair::getValue)
+              .map(v -> v.getRequest().toString())
+              .collect(Collectors.toList()));
+      // TODO restoreIndices?
```

Comment:
This is the case where the SQL commit was successful, otherwise we wouldn't be emitting MCLs. In this case one or more MCLs fail to be produced. We might try to recover this by running restoreIndices automatically, however it would also depend on producing messages to kafka and we'd end up in a loop. We are exposing this as a failure to the client, but they would have to call restoreIndices rather then retry to see any effect.